### PR TITLE
Added support for Inaccuracies for SpawnWeaponsWarhead

### LINKED ISSUE #452: `SpawnWeaponsWarhead`: add support for inaccuracy
Add support for inaccuracy, similar to how `CreateEffectWarhead` does it.

### PR DESCRIPTION
**Pull request by michal-mad**
Closes #452

### CODE REVIEW

**Comment by Dzierzan** [comment]
It's bugged, debris is not always spawned. In the GIF below I edited `Delays` and commented out `Inaccuracies`.

debris2

**Comment by Dzierzan** [comment]
Changes commited and squished. This should be ready for review @michaeldgg2 

There's however this issue where explosions do not follow the husk. I strongly believe this issue should be addressed in another PR.
wtp

**Comment by michaeldgg2** [comment]
Please rebase.

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 1]
```diff
@@ -20,6 +20,8 @@
 
 namespace OpenRA.Mods.OpenE2140.Warheads;
 
+using OpenRA.Effects;
```
Comment (IceReaper) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: Can we please keep all usings above the namespace? :)

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 2]
```diff
@@ -31,11 +33,14 @@ public class SpawnWeaponsWarhead : EffectWarhead, IRulesetLoaded<WeaponInfo>
 	public readonly int[] Delays = { 0 };
 
 	[Desc("The amount of projectile pieces to produce. Two values indicate a range.")]
-	public readonly int[] Pieces = { 3, 10 };
+	public readonly int[] Pieces = { 1, 1 };
```
Comment (IceReaper) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: Either the comment is wrong, or it should be a single value here: 1

Comment (michal-mad) [reply OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: I want for the default to be just 1 piece produced. Range is set from 1 to 1 - technically it's all good. Doing it this way makes the code cleaner (no checks for array boundary, single or double value etc).

It's OK it should stay this way.

Comment (michaeldgg2) [reply OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: There's `Util.RandomInRange` that you can use for this, which does all of this for you:

```
public static int RandomInRange(MersenneTwister random, int[] range)
{
	if (range.Length == 0)
		return 0;

	if (range.Length == 1)
		return range[0];

	return random.Next(range[0], range[1]);
}
```

**Comment by michal-mad** [comment]
Done

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 3]
```diff
@@ -68,62 +75,80 @@ public override void DoImpact(in Target target, WarheadArgs args)
...
 
 			for (var j = 0; pieces > j; j++)
 			{
 				var rotation = WRot.FromYaw(new WAngle(world.SharedRandom.Next(1024)));
+				var posCapture = pos;
 
 				var projectileArgs = new
 				{
 					Delay = this.Delays[i],
+					Inacurracy = inaccuracy,
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: `Inaccuracy` in this anonymous object is not used (inaccuracy is calculated into `pos` variable above.

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 4]
```diff
@@ -68,62 +75,80 @@ public override void DoImpact(in Target target, WarheadArgs args)
...
+			var pieces = Util.RandomInRange(world.SharedRandom, this.Pieces);
 			var range = world.SharedRandom.Next(this.Range[0].Length, this.Range[1].Length);
+			var inaccuracy = inaccuracies[i];
+
+			pos += WVec.FromPDF(world.SharedRandom, 2) * inaccuracy.Length / 1024;
 
 			for (var j = 0; pieces > j; j++)
 			{
 				var rotation = WRot.FromYaw(new WAngle(world.SharedRandom.Next(1024)));
+				var posCapture = pos;
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: Each piece's position needs to be random. Calculate it here and keep `pos` variable just for actor's center position.

**Review by michaeldgg2 — CHANGES_REQUESTED** [review]

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 5]
```diff
@@ -49,37 +54,47 @@ public void RulesetLoaded(Ruleset rules, WeaponInfo info)
...
+			? this.Inaccuracies 
+			: Enumerable.Repeat(this.Inaccuracies[0], this.WeaponInfos.Length).ToArray();
 
 		if (!this.IsValidAgainst(target, firedBy))
 			return;
 
 		for (var i = 0; this.WeaponInfos.Length > i; i++)
 		{
-			var pieces = world.SharedRandom.Next(this.Pieces[0], this.Pieces[1]);
+			var weaponImpactPosition = new WPos(target.CenterPosition.X, target.CenterPosition.Y, Math.Max(0, target.CenterPosition.Z));
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs:87]:
```suggestion
			var weaponImpactPosition = targetCenterPosition + WVec.FromPDF(world.SharedRandom, 2) * inaccuracy.Length / 1024;
```

There's no need to calculate target's center position in each iteration of outer loop.

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 6]
```diff
@@ -49,37 +54,47 @@ public void RulesetLoaded(Ruleset rules, WeaponInfo info)
...
+				return weapon;
+			})
 			.ToArray();
 	}
 
 	public override void DoImpact(in Target target, WarheadArgs args)
 	{
 		var firedBy = args.SourceActor;
-		var pos = new WPos(target.CenterPosition.X, target.CenterPosition.Y, Math.Max(0, target.CenterPosition.Z));
+		// var pos = new WPos(target.CenterPosition.X, target.CenterPosition.Y, Math.Max(0, target.CenterPosition.Z));
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs:76]:
```suggestion
		var targetCenterPosition = new WPos(target.CenterPosition.X, target.CenterPosition.Y, Math.Max(0, target.CenterPosition.Z));
```

**Review by michaeldgg2 — CHANGES_REQUESTED** [review]

**Comment by michaeldgg2** [comment]
Implemented in 44299bad58688ef8010270804e488e8ba750e236

**Pull request closed without merging**